ops/internal/deployer/deployer.go

```diff
@@ -0,0 +1,308 @@
+package deployer
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/BurntSushi/toml"
+	"github.com/ethereum-optimism/optimism/op-chain-ops/genesis"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// OpDeployer manages the process of building a specific binary of op-deployer,
+// then shelling out to that binary for various cli commands
+type OpDeployer struct {
+	DeployerVersion    string
+	binaryPath         string
+	lgr                log.Logger
+	l1ContractsRelease string
+}
+
+// NewOpDeployer creates a new OpDeployer instance.
+func NewOpDeployer(lgr log.Logger, l1ContractsRelease string) (*OpDeployer, error) {
+	if l1ContractsRelease == "" {
+		return nil, fmt.Errorf("l1ContractsRelease cannot be empty")
+	}
+
+	opd := OpDeployer{
+		lgr:                lgr,
+		l1ContractsRelease: l1ContractsRelease,
+	}
+
+	err := opd.checkBinary()
+	if err != nil {
+		return nil, fmt.Errorf("failed binary check: %w", err)
+	}
+
+	return &opd, nil
+}
+
+// checkBinary checks if the op-deployer binary exists and is executable
+func (d *OpDeployer) checkBinary() error {
+	// Normalize the contracts string before lookup in versions.json
+	// 1. Remove tag:// prefix if present
+	// 2. Remove any -rc.X suffix for version matching
+	contractsKey := strings.TrimPrefix(d.l1ContractsRelease, "tag://")
+	rcSuffixRegex := regexp.MustCompile(`-rc\.[0-9]+$`)
+	contractsKey = rcSuffixRegex.ReplaceAllString(contractsKey, "")
+
+	// Read and parse versions.json
+	d.lgr.Info("Reading versions.json")
+	// Get the directory containing this file (deployer.go)
+	_, filename, _, _ := runtime.Caller(0)
```

Comment:
If we absolutely must use versions.json, then let's at least embed it into the binary using `goembed` and unmarshal it once rather than reading it at runtime.

Reply:
Done in https://github.com/ethereum-optimism/superchain-registry/pull/1041/commits/1fba7e052bf9101c629c66b62fe5ee4e66b0c02d